### PR TITLE
[CPU] Enable a first (dummy) cache-level tiling for matmuls in codegen

### PR DESCRIPTION
**Pull request by dcaballe**
This PR is a step towards enabling a first cache-level tiling for matmuls. Currently, it introduces a new tiling level that replicates the tile sizes of the distribution level. This should be no-op in terms of performance as the single iteration loops should be eventually removed. This will also allow us to minimize the cache impact when changing the distribution sizes.

Currently, we are testing this in the codegen pass. To have this working, we have to switch the padding strategy to peeling, as padding has some implementation gaps that are exposed when we introduce more loops in the loop nest. There will be code size changes but that's the only way to move out of the local minima we are currently stuck upon. Consequently, this is removing the padding strategy from codegen.

We should be able to introduce cache-level tiling in a similar way for DT + UK in follow-up patches.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:834]: Do we want to use peeling expert? My understanding was that with masking we dont need peeling.

**Review by MaheshRavishankar — COMMENTED** [review]
I am all for not using padding pipeline. So thanks for doing that! Ill let @hanhanW review the details of the config settings, but I have only one comment as to why peeling expert and not just use masking?

**Comment by dcaballe** [comment]
> I am all for not using padding pipeline. So thanks for doing that! Ill let @hanhanW review the details of the config settings, but I have only one comment as to why peeling expert and not just use masking?

There are some masking issues when it comes to masking a scalar load (https://github.com/llvm/llvm-project/issues/63205) and the fact that can't be represented in LLVM. Quentin helped me with some of the changes needed in the backend and I have some patches to also land on the MLIR side but there's more work to be done. The MLIR patches were also blocked by supporting SSA values in `vector.extract`, which took much more than expected to land so I have to take that up again when priorities permit.

**Comment by MaheshRavishankar** [comment]
> > I am all for not using padding pipeline. So thanks for doing that! Ill let @hanhanW review the details of the config settings, but I have only one comment as to why peeling expert and not just use masking?
> 
> There are some masking issues when it comes to masking a scalar load ([llvm/llvm-project#63205](https://github.com/llvm/llvm-project/issues/63205)) and the fact that can't be represented in LLVM. Quentin helped me with some of the changes needed in the backend and I have some patches to also land on the MLIR side but there's more work to be done. The MLIR patches were also blocked by supporting SSA values in `vector.extract`, which took much more than expected to land so I have to take that up again when priorities permit.

Peeling does seem to increase dispatch size. I am trying to weight the advantage of getting rid of padding pipeline vs. that. If this was using masking for everything it would be a no brainer. Peeling has some impact on dispatch size.... 

Pulling in folks who might intersect with this to try this out. @qedawkins and @Max191 .

**Comment by dcaballe** [comment]
> Peeling does seem to increase dispatch size. I am trying to weight the advantage of getting rid of padding pipeline vs. that. If this was using masking for everything it would be a no brainer. Peeling has some impact on dispatch size....

We just need a way out of the rabbit hole we are currently in with padding. This is blocking the distribution and cache-level tiling work and I don't see a better alternative after looking into these for a few months. Peeling is the simplest and most robust approach that we have right now and we are using it already for some targets. This shouldn't have an impact if we are using UKs. However, we can disable it for any specific target where we have code size concerns.

**Comment by hanhanW** [comment]
I think we are able to drop padding pipeline once we turn on data-tiling by default. It will give us same (or more) coverage on matmul needs padding approach.

(I'm still having errands to run, I will come back to review the PR tomorrow.)

**Comment by hanhanW** [comment]
and yes, this is a cool work we will need for dt + uk as well!

**Comment by dcaballe** [comment]
Yes, this is for the non-DT side of things

**Comment by MaheshRavishankar** [comment]
Do you think it will be easier for landing this once the data tiling kicks in by default. I think that might help with the binary size (at least we then wont have to deal with peeling kicking in for contraction operations by default, which can have an impact on binary size.

**Comment by dcaballe** [comment]
If DT enablement becomes a blocker it would be difficult for us to make progress until then and we already have a few patches on the fly. Let's look at the data. This is only impacting x86. We shouldn't have code size issues there (large caches) and the numbers I see (92k -> 190k) seems pretty reasonable (i.e., it's not 10x). RISC-V is also testing this approach, with users pretty sensitive to code size, and we haven't seen issues either. The speedups on x86 also indicate that this is not harmful for performance so, in general, I don't see anything that indicates that landing this until we have DT is going to be a problem.

**Comment by MaheshRavishankar** [comment]
> If DT enablement becomes a blocker it would be difficult for us to make progress until then and we already have a few patches on the fly. Let's look at the data. This is only impacting x86. We shouldn't have code size issues there (large caches) and the numbers I see (92k -> 190k) seems pretty reasonable (i.e., it's not 10x). RISC-V is also testing this approach, with users pretty sensitive to code size, and we haven't seen issues either. The speedups on x86 also indicate that this is not harmful for performance so, in general, I don't see anything that indicates that landing this until we have DT is going to be a problem.

I think the IREE benchmarks CI gets only a weak signal of what happens. One of the issues I have seen with peeling is that for cases where vector size becomes big, you end up with a lot of instructions on two paths... I dont think DT enablement is too far off. @hanhanW thoughts?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:738]: Should we do an early-return if the strategy is VectorPreProcStrategy::None?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:785]: Where is `{8, 128, 16}` come from?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: TODO should be an actionable item and it's better to describe why this is TODO. Can you update the comment?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1103]: This can be simplified to a common form: `A ? A : B;`

```suggestion
    int64_t minTileSize = cacheTileSize ? cacheTileSize : vecTileSize;
```

**Comment by dcaballe** [comment]
I have also tested this on a few large models we are looking at, some with dynamic shapes, and still seeing no issues/perf improvements. There are a few correctness/perf issues that still need fixing on the DT side so it might take a little bit more.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1118]: Are we able to drop the if condition?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1154]: Is the TODO outdated?

**Review by hanhanW — CHANGES_REQUESTED** [review]

**Comment by hanhanW** [comment]
> I have also tested this on a few large models we are looking at, some with dynamic shapes, and still seeing no issues/perf improvements. There are a few correctness/perf issues that still need fixing on the DT side so it might take a little bit more.

Are there other pending PR based on the PR? If we can review them back-to-back, we wont sit in binary size regression for a long time.

This is not the only way to get out of the rabbit hole we are currently in with padding. DT enablement also addresses the issue. I think the remaining issue on your side is https://github.com/openxla/iree/issues/15061 and I need some time to finish my prototype. I'd appreciate if I can get some more help on the issue.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1154]: That's my TODO :) We've only enabled scalable sizes for non-padded matmuls, `setMatmulPeelingRootConfig()` does not yet setup scalable sizes in the `lowering_config`, so this is still a TODO.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1103]: Via a GCC extension you can do `A ?: B`  https://gcc.gnu.org/onlinedocs/gcc/Conditionals.html :P

Though personally I think the explicit `!= 0` check is a little clearer.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:785]: This is a value similar to the default vector sizes (actually let me use exactly that). These values are inputs to the computation of the distribution sizes so it really doesn't matter much right now.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:834]: Answered in the meeting. There are issues in the x86 backend.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1103]: I also like to make the int -> bool conversion explicit as I think it's easier to read. Thanks for the suggestion, though!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1118]: Not yet, unfortunately. We will have to remove all of this incrementally as most of the performance coming from padding is actually due to these numbers here, which seems to have some cache impact. So once, we decouple distribution from cache-level tiling we should be able to choose proper cache-level tile sizes for all the targets (or make them parametric, etc.)

**Review by dcaballe — COMMENTED** [review]
Thanks! Hopefully the comments provide more context.

**Review by hanhanW — APPROVED** [review]

**Comment by dcaballe** [comment]
I'll land this after rebasing and making sure there are no issue on CI and on internal models, hopefully by the end of the week, but, please, let me know if there are any other comments/concerns.

**Comment by MaheshRavishankar** [comment]
> I'll land this after rebasing and making sure there are no issue on CI and on internal models, hopefully by the end of the week, but, please, let me know if there are any other comments/concerns.

I wouldnt say it's a blocking concern, but I'd still suggest waiting for data tiling to be flipped to the default so that we don't need to use Peeling strategy. I am all for adding this multi level tiling (and I have used it recently and it's great to see it work pretty well), but peeling + matmul lowering hasn't worked well in the past. I realize that there is a stack of changes on top of this, but if data tiling is going to be flipped on by default soon, then the transience of using peeling for gemms seems unnecessary.

**Comment by hanhanW** [comment]
I don't have other concerns and generally agree with Mahesh. Matmul changes seems unnecessary because we will turn data-tiling on by default soon. But I'm open to keep a path for other approaches and other CPUs. Multi-level tiling is a good thing, and I will be happy to see it working!

**Comment by dcaballe** [comment]
I think we agreement was to land it and let it settle before we land anything else that depends on it. This has been waiting for 3 weeks already and it's really blocking the progress of cache-level tiling and distribution changes. If there is an issue that would lead to reverting this patch, I would like to know sooner than later but so far I haven't seen any major problems when testing this on CI and our large internal models. Peeling hasn't been problematic beyond increasing the code size and it's a sound strategy that is used in LLVM by default so DT enablement (which is something I'm also trying to enable) shouldn't be a blocker here.

**Comment by MaheshRavishankar** [comment]
> I think we agreement was to land it and let it settle before we land anything else that depends on it. This has been waiting for 3 weeks already and it's really blocking the progress of cache-level tiling and distribution changes. If there is an issue that would lead to reverting this patch, I would like to know sooner than later but so far I haven't seen any major problems when testing this on CI and our large internal models. Peeling hasn't been problematic beyond increasing the code size and it's a sound strategy that is used in LLVM by default so DT enablement (which is something I'm also trying to enable) shouldn't be a blocker here.

Sounds good! I am just stating my view for the record. I am not going to block it.

**Comment by dcaballe** [comment]
Bye, bye padding :)

**Pull request merged**
